### PR TITLE
PR: retriever interface

### PR DESCRIPTION
**Pull request by hbertrand**
Split the retriever from the documents manager. The goal is to have the read and write parts completely separated.

### CODE REVIEW

File: buster/busterbot.py [hunk 1]
```diff
@@ -64,11 +64,11 @@ class BusterConfig:
     source: str = ""
 
 
-from buster.documents.base import DocumentsManager
+from buster.retriever import Retriever
 
 
 class Buster:
-    def __init__(self, cfg: BusterConfig, documents: DocumentsManager):
+    def __init__(self, cfg: BusterConfig, documents: Retriever):
```
Comment (jerpint) [inline buster/busterbot.py:71]: maybe with the renaming we shoudl consider renaming `documents` to something else?

Comment (hbertrand) [reply buster/busterbot.py]: you're right, I renamed all instances to retriever

File: buster/retriever/pickle.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def __init__(self, filepath: str):
+        self.filepath = filepath
+        self.documents = pd.read_pickle(filepath)
+
+    def get_documents(self, source: str) -> pd.DataFrame:
+        if self.documents is None:
+            raise FileNotFoundError(f"No documents found at {self.filepath}. Are you sure this is the correct path?")
+
+        documents = self.documents.copy()
+        if "current" in documents.columns:
```
Comment (jerpint) [inline buster/retriever/pickle.py:16]: whats supposed to be in the `"current"` column? can we add a comment on that?

File: buster/utils.py [hunk 3]
```diff
@@ -34,3 +33,14 @@ def get_documents_manager_from_extension(filepath: str) -> Type[DocumentsManager
         return DocumentsDB
     else:
         raise ValueError(f"Unsupported format: {ext}.")
+
+
+def get_retriever_from_extension(filepath: str) -> Type[Retriever]:
```
Comment (jerpint) [inline buster/utils.py:38]: i dont think you need to explicitly mention `Type` here in the typehints, maybe just `-> Retriever` should be enough?

File: buster/retriever/pickle.py (hunk 2 above)
Comment (hbertrand) [reply buster/retriever/pickle.py:16]: just a 1 or 0, when there are multiple versions of a document. Mostly obsolete, but requires some change elsewhere to be properly removed, so it should go in another PR

File: buster/retriever/base.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+from openai.embeddings_utils import cosine_similarity
+
+
+@dataclass
+class Retriever(ABC):
+    @abstractmethod
+    def get_documents(self, source: str) -> pd.DataFrame:
+        ...
+
+    def retrieve(self, query_embedding: list[float], top_k: int, source: str = None) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (hbertrand) [reply buster/retriever/base.py:14]: As discussed on Slack, we need to evaluate our options first.

File: buster/retriever/pickle.py (hunk 2 above)
Comment (jerpint) [reply buster/retriever/pickle.py:17]: ok we should reference this in an issue somehow

Comment (jerpint) [reply buster/retriever/pickle.py:17]: referenced in #78

**Review by jerpint — APPROVED** [review]

**Pull request merged**